modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/TWKBReader.java

```diff
@@ -0,0 +1,442 @@
+        }
+    }
+
+    private MultiPolygon readMultiPolygon(TWKBMetadata metadata)
+            throws IOException, ParseException {
+        if (!metadata.isEmpty()) {
+            int numGeom = dis.readUnsignedInt();
+            Polygon[] geoms = new Polygon[numGeom];
+            for (int i = 0; i < numGeom; i++) {
+                geoms[i] = readPolygon(metadata);
+            }
+            return factory.createMultiPolygon(geoms);
+        } else {
+            return factory.createMultiPolygon();
+        }
+    }
+
+    private GeometryCollection readGeometryCollection(TWKBMetadata metadata)
+            throws IOException, ParseException {
+        if (!metadata.isEmpty()) {
+            int numGeom = dis.readUnsignedInt();
+            Geometry[] geoms = new Geometry[numGeom];
+            for (int i = 0; i < numGeom; i++) {
+                // from the spec, "for each geometry there will be a complete TWKB geometry,
+                // readable using the rules set out above", so need to reset the ordinates
+                if (i > 0) Arrays.fill(metadata.valueArray, 0);
+                geoms[i] = readGeometry();
+            }
+            return factory.createGeometryCollection(geoms);
+        } else {
+            return factory.createGeometryCollection();
+        }
+    }
+
+    private LinearRing readLinearRing(TWKBMetadata metadata) throws IOException {
+        int size = dis.readUnsignedInt(); // .readInt();
+        CoordinateSequence pts = readCoordinateSequenceRing(size, metadata);
+        return factory.createLinearRing(pts);
+    }
+
+    private CoordinateSequence readCoordinateSequenceRing(int size, TWKBMetadata metadata)
+            throws IOException {
+        CoordinateSequence seq = readCoordinateSequence(size, metadata);
+        return CoordinateSequences.ensureValidRing(csFactory, seq);
+    }
+
+    private int zigzagDecode(int input) {
+        return (input >> 1) ^ (-(input & 1));
+    }
+
+    private TWKBMetadata readMetadata() throws IOException {
+        int geometryTypeAndPrecision = dis.readByte();
+        int geometryType = geometryTypeAndPrecision & 0x0F;
+        int precision = zigzagDecode((geometryTypeAndPrecision & 0xF0) >> 4);
+
+        metadata.setType(geometryType);
+        metadata.setPrecision(precision);
+
+        byte header = dis.readByte();
+        metadata.setHeader(header);
+
+        int dims = 2;
+        if (metadata.hasExtendedDims()) {
+            int dimensions = dis.readByte();
+
+            if ((dimensions & 0x01) > 0) {
+                dims += 1;
+            }
+            if ((dimensions & 0x02) > 0) {
+                dims += 1;
+            }
+
+            if ((dimensions & 0x01) > 0) {
+                metadata.setHasM(true);
+                metadata.setMprecision((dimensions & 0xE0) >> 5);
+            }
+            if ((dimensions & 0x02) > 0) {
+                metadata.setHasZ(true);
+                metadata.setZprecision((dimensions & 0x1C) >> 2);
```

Comment:
It looks like these two ifs could be combined with the ones that increment the `dims` variable a few lines above